### PR TITLE
Updated x509 modules to match openssl 1.1.0h API

### PR DESCRIPTION
**Pull request by 1nfiniteloop**
This fixes issue #49

### CODE REVIEW

**Comment by 1nfiniteloop** [comment]
Ping, contributors! Could someone have a look at this pull request? The changes can be inspected in detail @ <https://github.com/1nfiniteloop/openssl/commits/x509-uplift>. Unittests in this library <https://github.com/1nfiniteloop/openssl-ca> might also help to verify this uplift.

**Comment by CyberShadow** [comment]
The diff is pretty big - the changes seem to be much more than just what was necessary to update the API to a newer version. It is also much too big to be reviewed by hand.

Could you go into some detail regarding the process used for creating this pull request?

Currently, this repository is in a bad shape. Different parts of the bindings have been updated using different methods to different OpenSSL API versions. Continuing to do this will further increase the fragmentation and inconsistency of the D bindings, so we probably should balance such attempts at partial improvements or at least make sure to document the process.

**Comment by 1nfiniteloop** [comment]
I decided to completely re-generate the x509 d-modules to avoid the inconcistency from previous partially updates and commits in the x509-modules, hopefully contribute shaping-up this library. This is indeed a big patch and a lot seems to be changed in the C api itself (commit e44c6af):

```
$ git show --stat e44c6af -- C/x509*
C/x509.h     | 1782 ++++++++++++++++++++++++++++++++++++++++++++------------------------------------------------------
C/x509_vfy.h |  841 +++++++++++++++++++++++-----------------------
C/x509v3.h   | 1269 +++++++++++++++++++++++++++++++++++-----------------------------------
```

The uplift is done with a combination of auto-generation and some manual patching. I agree that it needs documentation for describing how to upgrade. Hence I've created a draft guide from the steps I performed during the uplift. I'll add this document into my branch.

To summarize the workflow, the uplift has been done for each file as follows:

1. Generate d-module from C header (1.1.0h version) using d-step, commit file.
2. Inspect diff in header, Inspect diff in d-module, patch manually (according to check-list in guide). Make sure library builds, commit file.
3. Next file...

The six commits in branch <https://github.com/1nfiniteloop/openssl/commits/x509-uplift> has all these steps recorded as separate commits. When reviewing it might be the manual patching done after generation which is of most interest. Maybe we shouldn't merge the change squashed but rather have theses commits separate as a reference for future uplifts?

**Comment by CyberShadow** [comment]
> Hence I've created a draft guide from the steps I performed during the uplift. I'll add this document into my branch.

It would be good to have this documentation in the repository for future reference.

Could you also describe the methods you've used to verify the quality of the resulting bindings? E.g. I see that in the "Manual patching" commits, you added some declarations that were apparently missed from the automatically translated bindings. How did you find these and ensure that all instances of this error were thoroughly addressed?

**Comment by 1nfiniteloop** [comment]
The instructions is now available in <https://github.com/1nfiniteloop/openssl/blob/x509-uplift/doc/how-to-upgrade.md>. I have also added comments in the "Manual patching"-commits for review. I hope these will give answers on the questions.

**Comment by CyberShadow** [comment]
> The instructions is now available in https://github.com/1nfiniteloop/openssl/blob/x509-uplift/doc/how-to-upgrade.md.

Great, thanks. I think these should be part of this repository though, easily discoverable - so, either in or linked from `CONTRIBUTING.md`.

> I have also added comments in the "Manual patching"-commits for review. I hope these will give answers on the questions.

Great, but GitHub comments on commits are a bit ephemeral and not very discoverable. Ideally such documentation should be part of the git repository itself, either as comments in the source code, or in the commit messages. If you don't want to redo these comments, please add a link to the GitHub pages which contain them to the commit message here.

**Comment by 1nfiniteloop** [comment]
> Great, thanks. I think these should be part of this repository though, easily discoverable - so, either in or linked from CONTRIBUTING.md.

I added this guide into the other branch for reviewing purpose.  I consider to create a new branch and a new pull request _after_ this code-review with the fixes, and abandon this pull request to get a better git history for the merged content (Since this pull-request is a squash of all the changes). I believe it would be good to have the commits separated as a reference for future uplifts, rather than squashed.

I can change name to CONTRIBUTING.md if that's prefeerd?

> Great, but GitHub comments on commits are a bit ephemeral and not very discoverable. Ideally such documentation should be part of the git repository itself, either as comments in the source code, or in the commit messages. If you don't want to redo these comments, please add a link to the GitHub pages which contain them to the commit message here.

I added the comments to ease the code review. The information is already availalbe in the guide. I agree it could be clarified by adding comments in the commit messsages.

**Comment by CyberShadow** [comment]
>  I consider to create a new branch and a new pull request _after_ this code-review with the fixes

By "after" you mean after it's merged, or ...? After it's reviewed?

I personally can't review the diff directly (because of its size, and because I don't have any code using OpenSSL x509 features), so the best I can do is help review the methodology, to ensure that the changes are reproducible so that the same technique can be used for other parts of OpenSSL or future OpenSSL versions. I think it looks good from this point of view.

@carblue has been working on these bindings as well, so perhaps they may have something to say regarding this PR.

> I believe it would be good to have the commits separated as a reference for future uplifts, rather than squashed.

Agreed, so do you want to change this PR to use the non-squashed branch?

> I can change name to CONTRIBUTING.md if that's prefeerd?

Such documentation should be discoverable before someone wastes time figuring things documented there on their own, so I think that's a good idea.

**Comment by 1nfiniteloop** [comment]
>By "after" you mean after it's merged, or ...? After it's reviewed?

> Agreed, so do you want to change this PR to use the non-squashed branch?

After it's reviewed. Sorry to confuse you, I didn't realize that it's possible to edit an open pull request to point to another branch.

> Agreed, so do you want to change this PR to use the non-squashed branch?

Yes I will update this pull request to point to correct branch!

**Comment by carblue** [comment]
Sorry, I have no experience with OpenSSL's certificate/X509 API so far, nor code that uses that.
The binding that I'm preparing is based on  https://github.com/openssl/openssl, tag OpenSSL_1_1_1c, folder include/openssl. The X509* headers are still waiting to be translated, see https://github.com/carblue/openssl/blob/master/task_list_headers.txt; I'm working for those headers first that I already know to some extent.

As I'm outlining in README.md, the openssl 1.1.0 version will be unsupported by the OpenSSL maintainers in less than a month (no security fixes any more) and I believe there is a need to switch to openssl 1.1.1 soon (binary and an updated D binding)

**Comment by 1nfiniteloop** [comment]
>  ... the openssl 1.1.0 version will be unsupported by the OpenSSL maintainers in less than a month (no security fixes any more) and I believe there is a need to switch to openssl 1.1.1 soon (binary and an updated D binding)

Interesting, yes it would be good to upgrade the openssl-bindings to a recent and supported version! Sounds like a large amount of work? When looking into how much has changed between 1.1.0 and 1.1.1c example for this x509-module, it doesn't seem to be as comprehensive as this uplift though (see comment at the top). Uplifting to 1.1.1c will hopefully be smoother.

```
git diff --stat OpenSSL_1_1_0h OpenSSL_1_1_1c -- include/openssl/
...
 include/openssl/x509.h           |  148 ++------
 include/openssl/x509_vfy.h       |   86 ++++-
 include/openssl/x509err.h        |  125 +++++++
 include/openssl/x509v3.h         |  211 ++++-------
 include/openssl/x509v3err.h      |  158 +++++++++
```

Refeering to new pull request pointing to another branch with fixes in #54, closing this.
